### ISSUE TITLE
aarch64: user stack of child being modified after clone

### ISSUE THREAD

**Issue by CharlyYu** [QuarkContainer/Quark#1173]
The logs I added before and after clone indicates that the user stack of child is changed when I start nginx container on arm64.

```
@@ -423,6 +427,25 @@ pub fn SysClone(task: &mut Task, args: &SyscallArguments) -> Result<i64> {
         tls = args.arg3;
         cTid = args.arg4;
     }
+
+    let sp = task.GetPtRegs().get_stack_pointer();
+    let stack_bottom = Addr(sp).RoundUp()?.0;
+    let len = (stack_bottom - sp) as usize;
+    let ptr = sp as *const u8;
+    let buf = unsafe { slice::from_raw_parts(ptr, len) };
+    error!("Before clone, the sp is {:x}", sp);
+    let mut start = 0;
+    loop {
+        let mut end = start + 100;
+        if end > len {
+            end = len;
+        }
+        error!("buffer {}-{}: {:x?}", start, end, &buf[start..end]);
+        if end == len {
+            break;
+        }
+        start = end;
+    }

```

```

 #[inline]
 pub fn child_clone(userSp: u64) {
+    debug!("=========child clone: {:x}", userSp);
     let currTask = Task::Current();
     CPULocal::SetUserStack(userSp);
     CPULocal::SetKernelStack(currTask.GetKernelSp());

+    let stack_bottom = Addr(userSp).RoundUp().unwrap().0;
+    let len = (stack_bottom - userSp) as usize;
+    let ptr = userSp as *const u8;
+    let buf = unsafe { slice::from_raw_parts(ptr, len) };
+    error!("After clone, the sp is {:x}", userSp);
+    let mut start:usize = 0;
+    loop {
+        let mut end = start + 100;
+        if end > len {
+            end = len;
+        }
+        error!("buffer {}-{}: {:x?}", start, end, &buf[start..end]);
+        if end == len {
+            break;
+        }
+        start = end;
+    }
+
```

**Comment by CharlyYu** [comment]
Before clone:
```
[ERROR] [1/43c0140000|0] Before clone, the sp is 7fbcab25e7c0
[INFO] [1/772070] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 0-100: [0, e9, 25, ab, bc, 7f, 0, 0, 60, ab, b1, 4e, e8, d, 0, 0, d8, 8, b4, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a8, 8, b4, 4e, e8, d, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a0, e9, 25, ab, bc, 7f, 0, 0, f0, d0, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0]
[INFO] [1/772183] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 100-200: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, b0, e8, 25, ab, bc, 7f, 0, 0, ac, f0, a, 9b, 54, 2b, 0, 0, 0, 0, 1c, 9b, 54, 2b, 0, 0, 20, 6, 0, 9b, 54, 2b, 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 58, 0, 0, 0, 0, 0, 0, 0, 0, 60, 1c, 9b, 54, 2b, 0, 0, 78, 6, 0, 9b, 54, 2b, 0, 0, 7c, d3, b3, 4e, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, d0, f5, b3, 4e, e8, d, 0, 0]
[INFO] [1/772294] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 200-300: [a0, f7, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, f0, b3, 4e, e8, d, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, f0, e8, 25, ab, bc, 7f, 0, 0, c, b6, b1, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, 0, f0, b3, 4e, e8, d, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e]
[INFO] [1/772406] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 300-400: [e8, d, 0, 0, 0, e9, 25, ab, bc, 7f, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, 30, e9, 25, ab, bc, 7f, 0, 0, c0, 37, b1, 4e, e8, d, 0, 0, d8, 8, b4, 4e, e8, d, 0, 0, ff, ff, ff, ff, 0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, c0, e9, 25, ab, bc, 7f, 0, 0, 14, 34, b1, 4e, e8, d, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0]
[INFO] [1/772519] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 400-500: [2, 0, 0, 0, 0, 0, 0, 0, 10, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, d0, f5, b3, 4e, e8, d, 0, 0, a0, f7, b3, 4e, e8, d, 0, 0, c0, e9, 25, ab, bc, 7f, 0, 0, 18, 33, b1, 4e, 2, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0, 4, 0, 0, 0]
[INFO] [1/772632] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 500-600: [5, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, 30, ea, 25, ab, bc, 7f, 0, 0, 28, 35, b1, 4e, e8, d, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, 40, 6, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 18, 33, b1, 4e, e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0]
[INFO] [1/772745] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 600-700: [18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, a0, ea, 25, ab, bc, 7f, 0, 0, 14, 34, b1, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 40, 6, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, f0, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 18, 33, b1, 4e]
[INFO] [1/772861] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 700-800: [e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0, 18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, 10, eb, 25, ab, bc, 7f, 0, 0, d0, b2, b1, 4e, e8, d, 0, 0, 80, eb, 25, ab, bc, 7f, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ec, f5, b3, 4e, e8, d, 0, 0]
[INFO] [1/772995] hypecall: 8, paras: 43c015dc58 0 0 0
[ERROR] [1/43c0140000|0] buffer 800-900: [9c, d3, b3, 4e, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0, 18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, a0, eb, 25, ab, bc, 7f, 0, 0, 3c, 16, b1, 4e, e8, d, 0, 0, d0, f9, b3, 4e, e8, d, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 88, bb, b3, 4e, e8, d, 0, 0, 80, 14, b1, 4e, e8, d, 0, 0, a8, ed, 25, ab]
[INFO] [1/773110] hypecall: 8, paras: 43c015dc58 0 0 0

```

**Comment by CharlyYu** [comment]
[ERROR] [1/43c01a0000|0] After clone, the sp is 7fbcab25e7c0
[INFO] [1/1338813] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 0-100: [ff, ff, ff, ff, 0, 0, 0, 0, b0, e9, 25, ab, bc, 7f, 0, 0, 7c, d3, b3, 4e, e8, d, 0, 0, 48, 9, b4, 4e, e8, d, 0, 0, a0, e9, 25, ab, bc, 7f, 0, 0, f0, d0, b3, 4e, e8, d, 0, 0, a0, e9, 25, ab, bc, 7f, 0, 0, f8, 8, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, a0, e9, 25, ab, bc, 7f, 0, 0, f0, d0, b3, 4e, e8, d, 0, 0, e8, 8, b4, 4e]
[INFO] [1/1338926] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 100-200: [e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, ff, ff, ff, ff, 0, 0, 0, 0, b0, e9, 25, ab, bc, 7f, 0, 0, 7c, d3, b3, 4e, e8, d, 0, 0, 48, 9, b4, 4e, e8, d, 0, 0, a0, e9, 25, ab, bc, 7f, 0, 0, f0, d0, b3, 4e, e8, d, 0, 0, 1, 0, 0, 0, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, d0, f5, b3, 4e, e8, d, 0, 0]
[INFO] [1/1339041] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 200-300: [a0, f7, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4f, ef, 25, ab, bc, 7f, 0, 0, f8, 8, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, f0, e8, 25, ab, bc, 7f, 0, 0, c, b6, b1, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, 0, e9, 25, ab, bc, 7f, 0, 0, 80, ac, b1, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, 4, 0, 0, 0]
[INFO] [1/1339154] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 300-400: [0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0, 48, 9, b4, 4e, e8, d, 0, 0, 30, e9, 25, ab, bc, 7f, 0, 0, e4, 38, b1, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, e8, 8, b4, 4e, e8, d, 0, 0, d0, 18, b4, 4e, e8, d, 0, 0, c0, e9, 25, ab, bc, 7f, 0, 0, 14, 34, b1, 4e, e8, d, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, c8, 8, b4, 4e, e8, d, 0, 0]
[INFO] [1/1339268] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 400-500: [2, 0, 0, 0, 0, 0, 0, 0, 10, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, d0, f5, b3, 4e, e8, d, 0, 0, a0, f7, b3, 4e, e8, d, 0, 0, c0, e9, 25, ab, bc, 7f, 0, 0, 18, 33, b1, 4e, 2, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, d8, c6, b3, 4e, e8, d, 0, 0, 4, 0, 0, 0]
[INFO] [1/1339384] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 500-600: [ff, ff, ff, ff, 0, f9, ca, 86, e8, 97, f5, 48, 30, ea, 25, ab, bc, 7f, 0, 0, 28, 35, b1, 4e, e8, d, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, 40, 6, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 18, 33, b1, 4e, e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0]
[INFO] [1/1339506] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 600-700: [18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, a0, ea, 25, ab, bc, 7f, 0, 0, 14, 34, b1, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 40, 6, b4, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, f0, ea, 25, ab, bc, 7f, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, d0, b3, 4e, e8, d, 0, 0, 9c, d3, b3, 4e, e8, d, 0, 0, 18, 33, b1, 4e]
[INFO] [1/1339624] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 700-800: [e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0, 18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, 10, eb, 25, ab, bc, 7f, 0, 0, d0, b2, b1, 4e, e8, d, 0, 0, 80, eb, 25, ab, bc, 7f, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, a0, f9, b3, 4e, e8, d, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ec, f5, b3, 4e, e8, d, 0, 0]
[INFO] [1/1339779] hypecall: 8, paras: 43c01bf4b0 0 0 0
[ERROR] [1/43c01a0000|0] buffer 800-900: [9c, d3, b3, 4e, e8, d, 0, 0, 38, fa, b3, 4e, e8, d, 0, 0, 70, 16, b4, 4e, e8, d, 0, 0, 18, 17, b4, 4e, e8, d, 0, 0, 58, 1, 0, 0, 0, 0, 0, 0, 0, f9, ca, 86, e8, 97, f5, 48, a0, eb, 25, ab, bc, 7f, 0, 0, 3c, 16, b1, 4e, e8, d, 0, 0, d0, f9, b3, 4e, e8, d, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 88, bb, b3, 4e, e8, d, 0, 0, 80, 14, b1, 4e, e8, d, 0, 0, a8, ed, 25, ab]
[INFO] [1/1339904] hypecall: 8, paras: 43c01bf4b0 0 0 0

**Comment by shrik3** [comment]
I'll see what I can do.

**Comment by shrik3** [comment]
@CharlyYu 

Aside from the stack issue, it seems you are running both tasks on the same cpu?

As I tested, SMP scheduling doesn't seem to work ( see [this](https://github.com/QuarkContainer/Quark/pull/1106#issuecomment-2005007880) ). Have you tried a simple test fork program? Does it work on one CPU?

**Comment by CharlyYu** [comment]
It seems it is because the switch of ttbr0_el1 does not take effect instantly. some cache mechanism makes the memory translation with the old ttbr0_el1 value. I printed the values with the physical address of the stack, and it is ok, but with virtual address, I got the value of the old process' stack.

**Comment by shrik3** [comment]
> > Aside from the stack issue, it seems you are running both tasks on the same cpu?
> 
> Sorry I didn't mention that only 2 vcpus is allocated to the sandbox.

I'm interested if the simple fork program works for you? Because in my case it didn't work when two tasks are scheduled on the same vcpu.

**Comment by shrik3** [comment]
On Tue Mar 26, 2024 at 4:21 PM CET, CharlyYu wrote:
> It seems it is because the switch of ttbr0_el1 does not take effect instantly. some cache mechanism makes the memory translation with the old ttbr0_el1 value. I printed the values with the physical address of the stack, and it is ok, but with virtual address, I got the value of the old process' stack.


but we do have a tlb shootdown before returning to user process. Is is not
enough for cache consistency?

#[inline]
pub fn LoadUserTable(table: u64) {
    unsafe {
        asm!("msr ttbr0_el1, {0}
        ic iallu
        dsb nsh
        isb",
        in(reg) table);
    };
}

**Comment by CharlyYu** [comment]
I got it, I think we have to flush the tbl with tlbi, in x86_64, we just write the CR3 and the tlb will be flushed automatically, but in aarch64, we have to flush the local tlb all, I will submit a PR to fix it.

**Comment by CharlyYu** [comment]
And I found that many atomic operation is callled with memory ordering of Ordering::Relaxed, I think it also may not work for aarch64.

**Comment by CharlyYu** [comment]
Sorry I have to submit the pr tomorrow.

**Comment by shrik3** [comment]
> I got it, I think we have to flush the tbl with tlbi, in x86_64, we just write the CR3 and the tlb will be flushed automatically, but in aarch64, we have to flush the local tlb all, I will submit a PR to fix it.

Any update?  I can help if you are busy.

**Comment by chl337** [comment]
I think this is also addressed in the #1210. @CharlyYu can we close this one?

**Comment by shrik3** [comment]
#1212 should be the fix.
